File: documentation/AllMetaheuristic/others/CFO_re/CFO/CFO.m
% Central Force Optimization (CFO) Implementation


% Parameters
population_size = 50;    % Population size
num_variables = 10;      % Number of variables
max_iterations = 100;    % Maximum number of iterations
alpha = 0.1;              % Learning rate
lower_limit = -5;        % Lower limit for variables
upper_limit = 5;         % Upper limit for variables

% Initialization
population = lower_limit + (upper_limit - lower_limit) * rand(population_size, num_variables);  % Random initial population

% Main loop
for iter = 1:max_iterations
    % Evaluate objective function for each individual
    fitness = zeros(population_size, 1);
    for i = 1:population_size
        fitness(i) = objective_function(population(i, :));
    end
    
    % Sort population based on fitness
    [fitness, sorted_indices] = sort(fitness);
    population = population(sorted_indices, :);
    
    % Calculate center of mass
    center_of_mass = mean(population);
    
    % Update each individual's position
    for i = 1:population_size
        direction = center_of_mass - population(i, :);
        population(i, :) = population(i, :) + alpha * direction;
        
        % Ensure the new position is within bounds
        population(i, :) = max(min(population(i, :), upper_limit), lower_limit);
    end
end

% Find the best solution
[best_fitness, best_index] = min(fitness);
best_solution = population(best_index, :);
fprintf('Best solution: %s\n', mat2str(best_solution));
fprintf('Best fitness: %f\n', best_fitness);

% Objective function (example)
function z = objective_function(x)
    % Example: Sphere function
    z = sum(x.^2);
end

File: chekml/MetaheuristicOptimization/CIntegration_ctypes/CFO.c
/* CFO.c - Implementation file for Central Force Optimization */
#include "CFO.h"
#include "generaloptimizer.h"
#include <stdlib.h>
#include <time.h>

// Generate random double between min and max
double rand_double(double min, double max);

// Initialize Population
void initialize_population_cfo(Optimizer *opt) {
    for (int i = 0; i < opt->population_size; i++) {
        for (int j = 0; j < opt->dim; j++) {
            opt->population[i].position[j] = opt->bounds[2 * j] + 
                                           (opt->bounds[2 * j + 1] - opt->bounds[2 * j]) * rand_double(0.0, 1.0);
        }
        opt->population[i].fitness = INFINITY;
    }
    enforce_bound_constraints(opt);
}

// Central Force Update Phase
void central_force_update(Optimizer *opt) {
    double *center_of_mass = (double *)calloc(opt->dim, sizeof(double));
    
    // Calculate center of mass
    for (int i = 0; i < opt->population_size; i++) {
        for (int j = 0; j < opt->dim; j++) {
            center_of_mass[j] += opt->population[i].position[j];
        }
    }
    for (int j = 0; j < opt->dim; j++) {
        center_of_mass[j] /= opt->population_size;
    }

    // Update positions
    for (int i = 0; i < opt->population_size; i++) {
        for (int j = 0; j < opt->dim; j++) {
            double direction = center_of_mass[j] - opt->population[i].position[j];
            opt->population[i].position[j] += ALPHA * direction;
        }
    }

    free(center_of_mass);
    enforce_bound_constraints(opt);
}

// Update Best Solution
void update_best_solution_cfo(Optimizer *opt, double (*objective_function)(double *)) {
    for (int i = 0; i < opt->population_size; i++) {
        double new_fitness = objective_function(opt->population[i].position);
        if (new_fitness < opt->best_solution.fitness) {
            opt->best_solution.fitness = new_fitness;
            for (int j = 0; j < opt->dim; j++) {
                opt->best_solution.position[j] = opt->population[i].position[j];
            }
        }
    }
}

// Main Optimization Function
void CFO_optimize(Optimizer *opt, double (*objective_function)(double *)) {
    initialize_population_cfo(opt);
    
    for (int iter = 0; iter < opt->max_iter; iter++) {
        central_force_update(opt);
        update_best_solution_cfo(opt, objective_function);
        enforce_bound_constraints(opt);
        
        // Debugging output
        printf("Iteration %d: Best Value = %f\n", iter + 1, opt->best_solution.fitness);
    }
}